File: 2003/IntroToVBNETPart2/Presentation/CType.cs
using System ;

interface IAdd {
	int Add ( int x, int y ) ;
}

class Math : IAdd {
	int IAdd.Add ( int x, int y ) { return x + y ; }
}

public class TestCType {
	static void Main () {
		Object o = new Math () ;
		IAdd iAdd ;
		iAdd = (IAdd) o ;
		int z ;
		z = iAdd.Add ( 5, 6 ) ;
		Console.WriteLine ( "Sum of 5 + 6: {0}", z ) ;
		Math m ;
		// Compiler Error: Cannot implicitly convert type 'object' to 'Math'
		// m = o ;
		m = (Math) o ;
		Console.WriteLine ( "Sum of 5 + 6: {0}", ((IAdd)m).Add ( 5, 6 ) ) ;

		if  ( o  is ICloneable )
			Console.WriteLine ( "o is of type ICloneable" )  ;
		else
			Console.WriteLine ( "o is NOT of type ICloneable" ) ;
		if  ( o is IAdd  )
			Console.WriteLine ( "o is of type IAdd" ) ;
		else
			Console.WriteLine ( "o is NOT of type IAdd" )  ;
		if  ( o is Math ) 
			Console.WriteLine ( "o is of type Math" )  ;
		else
			Console.WriteLine ( "o is NOT of type Math" ) ;

		Math a ;
		IAdd b ;
		a = o as Math ;
		if ( a != null )
		{
			Console.WriteLine ( "Using as: Sum of 5 + 6: {0}", 
				( (IAdd) a).Add ( 5, 6 ) ) ;
		}
		// In this case, the as clause combines a test for an interface and conversion
		// into one statement, which is more efficient and provides better performance.
		b = o as IAdd ;
		if ( b != null )
		{
			Console.WriteLine ( "Sum of 5 + 6: {0}", b.Add ( 5, 6 ) ) ;
		}
	}
}

File: 2003/IntroToVBNETPart2/Presentation/CType.vb
Option Strict On
Imports System

Interface IAdd
    Function Add ( ByVal x As Integer, ByVal y As Integer ) As Integer
End Interface

Class Math
    Implements IAdd
    Public Function Add ( ByVal x As Integer, ByVal y As Integer ) As Integer _
            Implements IAdd.Add
        Return x + y
    End Function
End Class
Public Class TestCType
    Shared Sub Main ()
        Dim o As Object = New Math ()
        Dim iAdd As IAdd
        ' Option Strict On disallows implicit conversions from 'System.Object' to 'IAdd'
        ' iAdd = o 
        iAdd = CType ( o, IAdd )
        Dim z As Integer
        z = iAdd.Add ( 5, 6 )
        Console.WriteLine ( "Sum of 5 + 6: {0}", z )
        Dim m As Math
        ' Option Strict On disallows implicit conversions from 'System.Object' to 'Math'
        ' m = o
        m = CType ( o, Math )
        Console.WriteLine ( "Sum of 5 + 6: {0}", m.Add ( 5, 6 ) )

        If TypeOf ( o ) Is ICloneable Then
            Console.WriteLine ( "o is of type ICloneable" ) 
        Else 
            Console.WriteLine ( "o is NOT of type ICloneable" ) 
        End If
        If TypeOf ( o ) Is IAdd Then
            Console.WriteLine ( "o is of type IAdd" ) 
        Else
            Console.WriteLine ( "o is NOT of type IAdd" ) 
        End If
        If TypeOf ( o ) Is Math Then
            Console.WriteLine ( "o is of type Math" ) 
        Else
            Console.WriteLine ( "o is NOT of type Math" )
        End If
    End Sub 
End Class